### ISSUE TITLE
Vague error message for oneOfType

### ISSUE THREAD

**Issue by vinhlh** [facebook/prop-types#226]
### Type definitions
```js
    overlayClassName: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.shape({
        base: PropTypes.string.isRequired,
        afterOpen: PropTypes.string.isRequired,
        beforeClose: PropTypes.string.isRequired
      })
    ]),
```

### Usage
```js
overlayClassName={{
        base: 'modal-overlay',
        afterOpen: 'modal-overlay--opened'
      }}
```

### Error received
```
Warning: Failed prop type: Invalid prop `overlayClassName` supplied to `Modal`.
```

In my case, I am using react-modal, and after upgrading to new version, a new prop attribute is added.
It will be much better if we show some kinds of warnings:
```
Warning: Failed prop type: Invalid prop `overlayClassName.beforeClose` supplied to `Modal`. Expected `string` but  received `undefined`
```

I would like to improve it if you guys agree.

**Comment by neelav** [comment]
I'm running into this same issue.  It looks like the `validate` function inside `createUnionTypeChecker` is dropping the error message returned from `checker`.  Maybe it should concatenate the error messages together?

**Comment by ljharb** [comment]
Duplicate of #9.

**Comment by mplis** [comment]
@ljharb  #9 was closed by #198 but I'm still seeing this issue.

CodeSandbox: https://codesandbox.io/s/vague-error-message-for-proptypesoneoftype-v16ft

Code:

```jsx
function Button({ myProp, ...rest }) {
  return <button {...rest} />;
}

Button.propTypes = {
  myProp: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ])
};

function App() {
  return (
    <div className="App">
      <Button myProp={["hello"]}>Hello</Button>
    </div>
  );
}
```

PropTypes validation warning:

```
Warning: Failed prop type: Invalid prop `myProp` supplied to `Button`.
```

I would expect the warning to tell me that I need to pass in either a `string` or a `number` to `myProp`.

**Comment by ljharb** [comment]
@mplis that's because the [commit](https://github.com/facebook/prop-types/commit/3910873a2621a9f60a14d2ec5076a27e950e4eea) is currently unreleased.
